salt/orch/force-removal.sls

```diff
@@ -0,0 +1,130 @@
+set-cluster-wide-removal-grain:
+  salt.function:
+    - tgt: {{ all_up_nodes|join(',') }}
+    - tgt_type: list
+    - fail_minions: {{ all_up_nodes|join(',') }}
+    - name: grains.setval
+    - arg:
+      - force_removal_in_progress
+      - true
+
+# Try to sync some information before going on. We might use modules that have not been
+# discovered yet on the minions.
+sync-all:
+  salt.function:
+    - tgt: {{ all_up_nodes|join(',') }}
+    - tgt_type: list
+    - fail_minions: {{ all_up_nodes|join(',') }}
+    - names:
+      - saltutil.refresh_pillar
+      - saltutil.refresh_grains
+      - mine.update
+      - saltutil.sync_all
+
+{%- if target in all_up_nodes %}
+# Try to unregister the machine from cluster services and perform minimal housekeeping. It's
+# important to keep every sls in different toplevel stages, so each one will be tried even
+# if previous fails for whatever reason.
+{%- for service, sls_list in services_to_stop.items() %}
+  {%- for sls in sls_list %}
+stop-{{ service }}-{{ sls }}-stage:
+  salt.state:
+    - tgt: {{ target }}
+    - fail_minions: {{ target }}
+    - sls:
+        - {{ service }}.{{ sls }}
+  {%- endfor %}
+{%- endfor %}
+
+# Try to shutdown the target
+shutdown-target:
+  salt.function:
+    - tgt: {{ target }}
+    - name: cmd.run
+    - fail_minions: {{ target }}
+    - arg:
+      - systemctl poweroff
+    - kwarg:
+        bg: True
+{%- endif %}
+
+{%- if target in etcd_members %}
+# Try to remove the node from the etcd cluster from any other working machine.
+{%- set previous_node = none %}
+{%- for node in up_nodes_except_target %}
+try-to-unregister-etcd-{{ target }}-from-{{ node }}:
+  salt.function:
+    - tgt: {{ node }}
+    - name: caasp_etcd.member_remove
+    - fail_minions: {{ node }}
+    - kwarg:
+        nodename: {{ target_nodename }}
+  {%- if previous_node %}
+    - onfail:
+        - try-to-unregister-etcd-{{ target }}-from-{{ previous_node }}
+  {%- endif %}
+  {%- set previous_node = node %}
+{%- endfor %}
+{%- endif %}
+
+# Try to remove the node from the kubernetes cluster from any other working machine.
+{%- set previous_node = none %}
+{%- for node in up_nodes_except_target %}
+try-to-unregister-kubelet-{{ target }}-from-{{ node }}:
+  salt.function:
+    - tgt: {{ node }}
+    - name: cmd.run
+    - fail_minions: {{ node }}
+    - arg:
+        - kubectl --kubeconfig={{ pillar['paths']['kubeconfig'] }} delete node {{ target_nodename }}
+  {%- if previous_node %}
+    - onfail:
+        - try-to-unregister-kubelet-{{ target }}-from-{{ previous_node }}
+  {%- endif %}
+  {%- set previous_node = node %}
+{%- endfor %}
+
+# remove the we-are-removing-some-node grain in the cluster
+remove-cluster-wide-removal-grain:
+  salt.function:
+    - tgt: {{ all_up_nodes|join(',') }}
+    - tgt_type: list
+    - fail_minions: {{ all_up_nodes|join(',') }}
+    - name: grains.delval
+    - arg:
+      - force_removal_in_progress
+    - kwarg:
+        destructive: True
```

Comment:
According to https://docs.saltstack.com/en/latest/topics/orchestrate/orchestrate_runner.html#more-complex-orchestration: `Additionally, the states/functions will be executed in the order in which they are defined, unless prevented from doing so by any Requisites and Other Global State Arguments, as is the default in SLS files since 0.17.0`